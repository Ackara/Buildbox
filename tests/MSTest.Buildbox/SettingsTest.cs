using Acklann.Buildbox.SemVer;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Shouldly;
using System;
using System.IO;

namespace MSTest.Buildbox
{
    [TestClass]
    [UseApprovalSubdirectory(nameof(ApprovalTests))]
    [UseReporter(typeof(DiffReporter), typeof(ClipboardReporter))]
    public class SettingsTest
    {
        public TestContext TestContext { get; set; }

        [TestMethod]
        public void Load_should_generate_a_default_settings_file_when_it_do_not_exist()
        {
            // Act
            if (File.Exists(Settings.DefaultSettingsPath))
            {
                File.Delete(Settings.DefaultSettingsPath);
            }

            var result = Settings.Load();
            bool settingsFileWasCreated = File.Exists(Settings.DefaultSettingsPath);

            // Assert
            result.ShouldNotBeNull();
            settingsFileWasCreated.ShouldBeTrue();
            Approvals.VerifyFile(Settings.DefaultSettingsPath);
        }

        [TestMethod]
        public void Load_should_deserialize_a_settings_file_when_passed()
        {
            // Arrange
            string pathToConfigFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Samples", "semver", "config.json");

            // Act
            var results = Settings.Load(pathToConfigFile);

            // Assert
            results.ShouldNotBeNull();
            results.Version.Patch.ShouldBe(8);
            results.Version.Suffix.ShouldNotBeNullOrWhiteSpace();
        }

        [TestMethod]
        public void Load_should_deserialize_a_partial_settings_file_when_passed()
        {
            // Arrange
            string pathToSourceFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Samples", "semver", "config-partial.json");
            string pathToConfigFile = Path.Combine(TestContext.DeploymentDirectory, "config-partial.json");
            File.Copy(pathToSourceFile, pathToConfigFile, overwrite: true);

            // Act
            var settingsFile = Settings.Load(pathToConfigFile);
            var contentsBeforeSave = File.ReadAllText(pathToConfigFile);
            settingsFile.Save(partial: true);
            var contentsAfterSave = File.ReadAllText(pathToConfigFile);

            // Assert
            settingsFile.ShouldNotBeNull();
            Approvals.VerifyJson(contentsAfterSave);
            contentsBeforeSave.ShouldBe(contentsAfterSave);
        }
    }
}