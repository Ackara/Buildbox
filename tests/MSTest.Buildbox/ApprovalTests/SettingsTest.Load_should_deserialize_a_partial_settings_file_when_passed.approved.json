{
    
  "metadata": 
  {
        
    "owner":  "Ackara",
        
    "projectUrl":  ""
  
    },
    
  "nested": 
  {
        
    "semver": 
    {
            
      "version": 
      {
                
        "major":  4,
                
        "minor":  0,
                
        "patch":  8,
                
        "suffix":  "-alpha"
      
            },
            
      "branchSuffixMap": 
      {
                
        "master":  "",
                
        "*":  "alpha"
      
            },
            
      "shouldCommitChanges":  true,
            
      "shouldTagCommit":  false,
            
      "shouldAddAllUnstagedFilesWhenCommitting":  false
    
        }
  
    }

}